core/src/main/java/org/elasticsearch/index/IndexSortConfig.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index;
+
+import org.apache.lucene.search.Sort;
+import org.apache.lucene.search.SortField;
+import org.apache.lucene.search.SortedNumericSortField;
+import org.apache.lucene.search.SortedSetSortField;
+import org.elasticsearch.Version;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.search.MultiValueMode;
+import org.elasticsearch.search.sort.SortOrder;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * Holds all the information that is used to build the sort order of an index.
+ *
+ * The index sort settings are <b>final</b> and can be defined only at index creation.
+ */
+public final class IndexSortConfig {
+    /**
+     * The list of field names
+     */
+    public static final Setting<List<String>> INDEX_SORT_FIELD_SETTING =
+        Setting.listSetting("index.sort.field", Collections.emptyList(),
+            Function.identity(), Setting.Property.IndexScope, Setting.Property.Final);
+
+    /**
+     * The {@link SortOrder} for each specified sort field (ie. <b>asc</b> or <b>desc</b>).
+     */
+    public static final Setting<List<SortOrder>> INDEX_SORT_ORDER_SETTING =
+        Setting.listSetting("index.sort.order", Collections.emptyList(),
+            IndexSortConfig::parseOrderMode, Setting.Property.IndexScope, Setting.Property.Final);
+
+    private static SortOrder parseOrderMode(String value) {
```

Comment:
I'd probably move these private methods below all the public ones, even if they are static.